### ISSUE TITLE
What should a Pattern contain?

### ISSUE THREAD

**Issue by inf3rno** [inf3rno/patterns#3]
Iteration from #2 
```js
match = helper.find(string, fromIndex, "(\\()|(\\w)|(\\))", ["parameters.open", "parameter", "parameters.close"]);
match.index; // 0
match.name; // "parameters.open"
match.range; // [firstPosition, lastPosition]
match.groups; // {}
match.substring; // "("
```

**Comment by inf3rno** [comment]
What about named capturing groups?

In theory they are supported by ES2018 https://stackoverflow.com/questions/5367369/named-capturing-groups-in-javascript-regex For older environments we need to map indexed groups with them.

**Comment by inf3rno** [comment]
I am curious whether indexed groups can be easily parsed and mapped. In theory I can do `[(?<name>)]` and as far as I can tell with my regex knowledge that is a completely valid character class. So it is not enough to replace the `(?<name>value)` with `(value)` in the pattern. We need to parse the regex pattern to do this properly. So even named capturing groups should be built on the upper iterator, which means an infinite loop, because the iterator should depend on named capturing groups too... What we need is a low level iterator, that does not do index - name mapping or does not use named capturing groups.

**Comment by inf3rno** [comment]
```js
var lexer = createLexer({
	parameters: "()".split(""),
	parameter: latin1.word
});
// ->
/(?<parameters_open>\()|(?<parameter>\w+)|(?<parameters_close>\))/g
```

I am not sure what a capturing group name can contain, so even if these names are supported, we might need to do mapping if we want to use utf-8 characters.

**Comment by inf3rno** [comment]
Hmm I might need https://github.com/slevithan/xregexp I have to check that lib.

**Comment by inf3rno** [comment]
I guess I already have an answer: https://github.com/slevithan/xregexp/issues/224 The capturing group names are variable identifiers instead of \w+, so I need to support the feature that way.

**Comment by inf3rno** [comment]
I won't use this lib because I'd rather generate the regex pattern instead of writing it, so the usage of mapping vs named capturing groups won't matter here.

```js
or(capture("parameters.open", "("), capture("parameter", word()), capture("parameters.close", ")"))
/(?<parameters_open>\()|(?<parameter>\w+)|(?<parameters_close>\))/g
```

**Comment by inf3rno** [comment]
What I really hate here, that as far as I remember if the parameter cg matches, the parameters_open cg will contains empty string, while the parameters_close will contain null or undefined. I need to check it though. I need to know in these cases which capturing group was matching and which wasn't. I think these are special cases for the regex based parser, but I might have to think about more general solutions in the topic. I really want to know which group was matching, because the lexer needs to add the token type to the result.

**Comment by inf3rno** [comment]
Well I don't really want to care about generating utf-8 word patterns, describing quantifiers, etc. Maybe another time. For now it is more than enough to use a hybrid approach like this:
```js
var pattern = new Pattern(or(
    capture("parameters.open", "("), 
    capture("parameter", regex("\\w+")), 
    capture("parameters.close", ")")
));
```

We could build a templating language to make this shorter, but I never liked inventhing a new language to solve a problem, so I'd rather use js with implicit rules e.g. an array means "or" or something like that. I am not sure yet.

```js
var pattern = new Pattern({
    "pattern": "${parameters.open}|${parameter}|${parameters.close}",
    "parameters.open": "(", 
    "parameter": regex("\\w+"), 
    "parameters.close": ")"
});
```

**Comment by inf3rno** [comment]
Another problem is the pointer.
```js
var match = pattern.find(string, 0);
```
This translates to
```js
regex.lastIndex = 0;
var match = regex.exec(string);
```
I need to track the string - pointer pairs by using a pattern. Doing `string: pointer` maps is out of the question, because as far as I know it would be very slow, because it uses the value as a key instead of the memory address.
Wrapping the string is the only solution can think of.
```js
var pointer = new Pointer(string, 0);
var match1 = pattern.find(pointer);
var match2 = pattern.find(pointer);
var match3 = pattern.find(pointer);
// ...
```
If I want to sacrifise immutability ofc.

Another option to do it with immutable pointer, but I find this disgusting:

```js
var pointer0 = new Pointer(string, 0);
var {match: match1, pointer: pointer1} = pattern.find(pointer0);
var {match: match2, pointer: pointer2} = pattern.find(pointer1);
var {match: match3, pointer: pointer3} = pattern.find(pointer2);
// ...
```

**Comment by inf3rno** [comment]
Immutability does not make sense if we wrap both the string and the lastIndex, so either we keep the current regex approach and do something like `pattern.postMatchIndex` or add it to the match as `match.toIndex + 1` or we make it mutable.

**Comment by inf3rno** [comment]
I don't want to use `pattern.lastMatchPointer`. It is a memory leak, because it does not allow the GC to destroy the string we used with the pattern until we use another string. The `pattern.postMatchIndex` is ok, because it does not take much memory.

**Comment by inf3rno** [comment]
Returning `{match,pointer}` is out of the question. The maximum I would do is `match = {pointer, groups}`. If we mutate the match by overriding its value, then we can do this:

```js
var results = [];
var pointer = new Pointer(string, 0);
var match = pattern.find(pointer);
results.push(match.groups);
var match = pattern.find(match.pointer);
results.push(match.groups);
var match = pattern.find(match.pointer);
results.push(match.groups);
```

**Comment by inf3rno** [comment]
How should the match look like?

The current match is the following: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/RegExp/exec
I am not sure how Chrome solved it with named capturing groups. Tbh. I don't really care. It gives some sort of array where 0 is the substring, 1 is the first capturing group, and so on. It has an input property, which stores the original string. For me the pointer will be good for that. It has an index property, which stores the first index of the match.

So it needs to contain:
 - first index + substring.length or index range of the match
 I would go for `{fromIndex,toIndex}` or `{firstIndex, lastIndex}` even though the lastIndex is taken by the regex and has a different meaning.
 - matching substring
 Which can be `matchingSubstring` or `substring`.
 - capturing groups
 Which can be `groups`, `capturingGroups`.

So a possible solution is this:

```js
var match = {substring, groups, fromIndex, toIndex};
```

While this is all nice and clear and I am not fond of hybrid objects, I'd like to check the XRegExp and the normal regex solution where they mix the capturing groups with the upper properties. According to XRegExp it has an advantage. At least they are doing it that way.

```js
const date = XRegExp(
    `(?<year>  [0-9]{4} ) -?  # year
     (?<month> [0-9]{2} ) -?  # month
     (?<day>   [0-9]{2} )     # day`, 'x');
let match = XRegExp.exec('2017-02-22', date);
match.year;
```

**Comment by inf3rno** [comment]
I am not sure I want to keep the lastIndex/toIndex. I mean I can count it with  `firstIndex + substring.length -1` or with `pointer.index -1`. I am sure I won't use the lastIndex name, because it would lead to confusion. I don't like the fromIndex and toIndex that much, but they are acceptable. I might stick with only an index variable. `{groups,index,substring}`, where the groups are mapped to property names.

**Comment by inf3rno** [comment]
What about having unnamed groups?
I could do `groups[index] = value` or I could simply forget about them.

**Comment by inf3rno** [comment]
How to handle nested capturing groups?

I hate this flattened approach by naming them. I would rather use nested objects.
```
(a(b(c))) -[a,b,c]-> {a:{b:"c"}}
```
What is really bad here that the values of the branch groups are not stored in the result. From that perspective having a non-enumerable `substring` property would be good after all:

```
{
    substring: "abc",
    a: {
        substring: "bc",
        b:"c"
    }
}
```

I just need to decide which environments I want to support. Enumerable properties are ES5 features I think. ES6 added Symbols which would be even better here.

**Comment by inf3rno** [comment]
I'd like the symbols, but I want this to be ES5 compatible. I'd rather use a prefix like @ if necessary. Or just make index and substring reserved words. We can add "matching" as prefix. So it would be `matchingSubstring` and `matchingIndex`.

**Comment by inf3rno** [comment]
I never liked the string replacements, so instead of using templates there too `${a.b.c}` I'd like to pass the match object to the function.

**Comment by inf3rno** [comment]
I have to rethink the groups: `(a(b(c))) -[a,b,c]-> {a:{b:"c"}}`. Let's reduce it.

```
abc -> {substring: "abc"}
(abc) -> {substring: "abc", a: "abc"}
(a(bc)) -> {substring: "abc", a: {substring:"abc", b: "bc"}}
(a(b(c))) -> {substring: "abc", a: {substring: "abc, b: {substring: "bc", c: "c"}}}
# or without the substrings:
(a(b(c))) -> {a: {b: {c: "c"}}}
```

So the right syntax is:
`(a(b(c))) -[a,b,c]-> {a:{b:{c:"c"}}}`

**Comment by inf3rno** [comment]
For the substring we can have a relative good implicit rule. We can use toString() to write the substring and maybe valueOf to write the index on the match.

**Comment by inf3rno** [comment]
For the pattern creation I made up my mind about the template. It would be nice if I can keep it simple. So the maximum it should do is adding a variable. Function calls, iterations, etc. would be too much. Let's use the date example:

```js
const date = new Pattern({
    main: "${year}-${month}-${day}",
    year: "[0-9]{4}",
    month: "[0-9]{2}",
    day: "[0-9]{2}",
});
let match = date.find('2017-02-22');
match.year;
```

**Comment by inf3rno** [comment]
Possible solutions for group separations:

```js
{
    main: "(${a})-(${b})",
    a: "${positiveInteger}",
    b: "${positiveInteger}",
    positiveInteger: "\\d+"
}
```

```js
{
    main: "${(a)}-${(b)})",
    a: "${positiveInteger}",
    b: "${positiveInteger}",
    positiveInteger: "\\d+"
}
```

```js
{
    main: "${a}-${b}",
    "(a)": "${positiveInteger}",
    "(b)": "${positiveInteger}",
    positiveInteger: "\\d+"
}
```

```js
{
    main: "${a}-${b}",
    a: "@{positiveInteger}",
    b: "@{positiveInteger}",
    positiveInteger: "\\d+"
}
```

```js
{
    main: "${a}-${b}",
    a: "${_positiveInteger}",
    b: "${_positiveInteger}",
    _positiveInteger: "\\d+"
}
```

**Comment by inf3rno** [comment]
I like the 1,3. I hate the 2,5. I am not sure about the 4.

**Comment by inf3rno** [comment]
I am sure I want to limit capturing group usage in regex, so I will partially parse the pattern and throw error if I find a capturing group in it. Another option for the template:

```js
{
    main: "(?<a>)-(?<b>)",
    a: "(?<positiveInteger>)",
    b: "(?<positiveInteger>)",
    positiveInteger: "\\d+"
}
```

**Comment by inf3rno** [comment]
Another idea here that we could use the nested syntax instead of the flattened, but it would not work if we want to allow reusing non-capturing groups. Even the upper one is a graph rather than a tree.

**Comment by inf3rno** [comment]
Yet another option is listing what we want to capture.
```js
var pattern = new Pattern({
    "*": "${a}-${b}",
    "a": "${positiveInteger}",
    "b": "${positiveInteger}",
    positiveInteger: "\\d+"
});
pattern.find("19-23", {a:{positiveInteger:1}, b: {positiveInteger: 1}});
```
Actually here we can use a tree by typing the positiveInteger twice, and if we are talking about the results, they could work the same way even if we are talking about graphs. That means we can reuse the capturing groups too, not just the non-capturing ones. We can go even further:

```js
var pattern = new Pattern({
    "*": "${positiveInteger}-${positiveInteger}",
    positiveInteger: "\\d+"
});
pattern.find("12-23"); // {positiveInteger: ["12","23"]}
```

**Comment by inf3rno** [comment]
We need to decide on which level we want to decide about capturing. Let's assume we have a pattern like `ąbc`, where "a","b","c" are subpatterns. How should we capture "b"?

1. capture by subpattern usage
```js
{"*": "${a}${capture(b)}${c}", a: "", b: "...", c: ""}
```
2. capture by subpattern definition in the key
```js
{"*": "${a}${b}${c}", a: "", "capture(b)": "...", c: ""}
```
3. capture by subpattern definition in the value
```js
{"*": "${a}${b}${c}", a: "", b: "capture(...)", c: ""}
```
4. capture all and select what we need later
```js
{"*": "${a}${b}${c}", a: "", b: "...", c: ""}
```

**Comment by inf3rno** [comment]
Capture by usage is interesting, because we can give a different name to the group as we have by the variable.

```js
{"*": "${a}${b->x}${c}", a: "", b: "...", c: ""}
```

Another option for the same is adding a mapping and if so I'd rather use the same concept as we got by destructuring:

```js
pattern = {"*": "${a}${b}${c}", a: "", b: "...", c: ""}
mapping = {b: "x"}
```

Yet another option is using destructuring and call a function by every pattern match:

```js
pattern = {"*": "${a}${b}${c}", a: "", b: "...", c: ""}
pattern.match(pointer, function ({b: x}){});
```

**Comment by inf3rno** [comment]
For the parser what I need is `${a}|${b}|${c}` and capturing all a,b,c. These patterns exclude each other, so it would be nice to have a feature to know which one matched and which did not. We have an analogy here to the single select input in HTML. We have different alternatives and we can select only one. We might be able to use the same model. The `or(a,b,c)` was good.

Using a constructor function or real variables might be just as good as using a template. We just need to develop it further:
```js
var {or,capture,word} = require("patterns");
var pattern = new Pattern(or(
    capture("parameters.open", "("), 
    capture("parameter", word()), 
    capture("parameters.close", ")")
));
var {parameter} = pattern.find(string, 0);
```

**Comment by inf3rno** [comment]
```js
var pattern = (function (){}(
	var a = new CapturingGroup("a", "text");
	var b = new CapturingGroup("b", new NumberPattern());
	var c = new CapturingGroup("c", new WordPattern(10,15));
	return new Pattern(new AndOperator(a,b,c));
))();
```
What I really hate here that I cannot use the variable names as capturing group names. I guess that's why I use an object on the template, so I don't have to type the name twice.

**Comment by inf3rno** [comment]
```js
var a = quote("text");
var b = number();
var c = word(10,15);
var pattern = and(a,b,c);
```
mapping
1.
```js
pattern.map({x: a});
```
2.
```js
pattern.map({"1":"x"});
```
3.
```js
pattern.map([x,a]);
```

This is another option. We let the library decide about capturing groups by mapping. The only problem is reusability here. It is the same as "capture by subpattern definition in the key" in properties, so we bind our mapping to the definition and my reusing the subpattern we will capture all its usages. If we can solve this in a simple more or less explicit way, then I will prefer this approach over using templates, because it is flexible and uses capturing groups on demand. The only problem that we need to compile the regex whenever the groups change, but I can live with that, and they won't change frequently...

**Comment by inf3rno** [comment]
Mutex is not entirely what I was talking about here. It is about choosing from alternatives, while mutex is about locking a resource. https://en.wikipedia.org/wiki/Mutual_exclusion So it is not a good name, because it has a slightly different meaning in a different context.

**Comment by inf3rno** [comment]
The `or` is fine, but I don't like the `and` here. Because it means both at once, not concatenation. And if I use `concat(enate)`, then I won't use `or`. I rather use `alt(ernatives)` or `opt(ions)`. The `choose`, `choice`, `select(one)` are not good enough.

**Comment by inf3rno** [comment]
Let's check the backreference. We have a text like `"13abc13"` and a pattern like `(\d+)(\w+)\1`. We need to be able to generate this pattern without any issue.

**Comment by inf3rno** [comment]
```js
var a = capture(number());
var b = capture(word());
var pattern = concat(a,b,backreference(a));
```
In this case it is not a problem.

**Comment by inf3rno** [comment]
We have two options here. We create and name the capturing groups by variable usage. E.g.
```js
var a = number();
var b = word();
var pattern = concat(capture(a).as("a"),capture(b).as("b"),backreference("a"));
```

Or we restrict the capturing group usage and allow only a single usage of the same capturing group.

Or maybe another solution is allowing multiple usage and decide based on logic which capturing group to backreferene. e.g. in the case of `or(concat(a,c,ref(a)), concat(a,b,ref(a)))` it is easy to determine the result, because the `or()` makes the usage of the first `a` impossible. On the other hand if we do `concat(a,a,ref(a))`, then we have a problem...

**Comment by inf3rno** [comment]
So the final solution is allowing the usage of a single capturing group only once. We can make an exception when it is evident which usage we want to backreference. Another option is this:

```js
var a = capture(number());
var b = clone(a);
var p = concat(a,b,a);
```
In this case the second usage can be backreference. So we won't need to use `ref()` which makes it simpler, but the rule is implicit here... Better to automatically clone the capturing group in this case or throw error because we used it twice.

**Comment by inf3rno** [comment]
I think I'll start with a strict approach. So a single capturing group can be used only once. I might loosen it later. I won't support `clone` for now. It is possible to solve everything without it.

**Comment by inf3rno** [comment]
I am curious if there is an oop solution for this. Using multiple functions make this hard to use.

```js
var a = capture(number());
var b = clone(a);
var p = concat(a,b,ref(a));
```
vs
```js
var a = patterns.number().capture();
var b = a.clone();
var p = a.concat(b).concat(a.reference());
```

**Comment by inf3rno** [comment]
Another interesting question if we can make capture automatical.

```js
var a = capture(number());
var b = clone(a);
var p = concat(a,b,ref(a));
var map = {x:a,y:{z:b}};
var pattern = pattern(p, map);
pattern.find(pointer("134513",0)); // {x:13, y:{z:45}}
```
vs

```js
var a = number();
var b = clone(a);
var p = concat(a,b,ref(a));
var map = {x:a,y:{z:b}};
var pattern = pattern(p, map);
pattern.find(pointer("134513",0)); // {x:13, y:{z:45}}
```

In the latter case we still have to add variables for capturing groups, but we don't have to call "capture" and it still makes sense... The only difference here, that every sub-pattern has to get a unique identifier. But we can use lazy generation for that if it is needed.

We can add strings for one liners:

```js
var pattern = pattern(concat(number().as("a"),clone("a"),ref("a")), {x:"a",y:{z:"b"}});
pattern.find(pointer("134513",0)); // {x:13, y:{z:45}}
```

**Comment by inf3rno** [comment]
What do we need for basic usage?

for pattern generation:
- capturing groups
- backreference
- character classes
- quantifiers
- primitives: 
  - text/string/quote/escape, 
  - number(integer,float,signed?)
  - date?, time? - maybe later, it is too complex to contain by default
  - word - latin1?, utf8
  - range? (generating patterns for number range?)
- begin/end/endOfLine
- uc,lc
- lookahead?, lookbehind?
- recursion?

for usage:
- pattern compilation with a map
- single match on the whole text
- multiple matches on the whole text with gaps between them
  - only patterns in iteration
  - both gaps and patterns in iteration
- multiple matches covering the whole text

**Comment by inf3rno** [comment]
compilation:
```js
var pattern = pattern(p, map);
```
I don't like the "pattern" keyword here. I would rather use "compile":

```js
var pattern = compile(p, map);
```

**Comment by inf3rno** [comment]
Hmm we could use the "index", "lastIndex", etc. in the map too.
 ```js
var pattern = compile(quote("abc"), {index: "@index"});
```

**Comment by inf3rno** [comment]
I am not sure how this is compatible with our `toString` approach by nested capturing groups.
let's check it on pattern: `(a(b(c)))`
the result should be `{toString: ()=>"abc", a:{toString: ()=>"abc", b: {toString: ()=> "bc", c: "c"}}}`
How should the pattern generation and the map look like?

**Comment by inf3rno** [comment]
The map could be:
```js
{
    "@toString": "@match", 
    a: {
        "@toString": a,
        b: {
            "@toString": b,
            c: c
        }
    }
}
```

**Comment by inf3rno** [comment]
Beside of iteration we will need a `findAll` or `matchAll` method too. I like this mapping approach. It adds capturing groups by usage and still it is simple and explicit enough. It won't be auto generated except if we allow the `.as("group")` for auto generation. For now I certainly won't allow it and I think I won't allow it later either.

**Comment by inf3rno** [comment]
I need a simple way to decide by `alternatives` which one matched. I don't like the `match.x === undefined` approach. There must be a better way.

**Comment by inf3rno** [comment]
This is the normal regex strategy:
```js
var a = number();
var b = word();
var x = alternatives(a,b);
var p = compile(x, {a:a,b:b});
p.find(pointer("abc",0));
// {a:undefined, b: "abc"}
```

The regex looks like this: `(?:(\d+)|(\w+))`.

I think we should connect the mapping to the alternatives somehow. We might need a capturing group for the alternatives too:  `((\d+)|(\w+))`, but it is possible to solve without that too.

We need a different mapping.
```js
var a = number();
var b = word();
var x = alternatives(a,b);
var p = compile(x, {x:{"@base": x, "@alternatives": {a:a,b:b}}});
p.find(pointer("abc",0));
// {x: {"@type": "b", "@value": "abc"}}
```

It is ugly, but works. Probably I can make it better.

**Comment by inf3rno** [comment]
I think we should use "@alternative" instead of "@type". I am not sure about the value. It could be `b: "abc"` instead and we could use the "@alternative" as key to get the value. We don't really need the branch "x" here.

```js
var a = number();
var b = word();
var x = alternatives(a,b);
var p = compile(x, {x:x,y:alternative({a:a,b:b}),a:a,b:b});
p.find(pointer("abc",0));
// {x:"abc", y:"b", a:undefined, b:"abc"}
```

I did not think of that, but instead of special properties with the "@" prefix we could call functions to build the mapping too...

I am not sure if this contains enough info: `y: alternative({a:a,b:b})`. It makes implicit that we are talking about the "x" alternatives. So if we use these subpatterns in multiple alternatives, then we have a problem.

```js
var a = number();
var b = word();
var x = alternatives(a,b);
var y = alternatives(a,b,);
var mapping = "?";
var p = compile(concatenate(x,y), mapping);
p.find(pointer("abc123",0));
// {x:"b",y:"a"}
```

In this case it is important to add the mapping that we need "x" or "y" alternatives. From other perspective it is completely fine to reuse the sub-patterns I guess.

`var mapping = {x: alternative(x, {a:a,b:b}), y: alternative(y, {a:a, b:b})};`

Instead of the "alternative" would could use "selection", "match", "choice", etc. For me the "selection" looks good enough, but it is not that good in thise context. It was nice by the select input in HTML, where we selected from alternatives (options) too. Here we rather match the alternative than select it.

**Comment by inf3rno** [comment]
I still don't like the "alternative", because of the abbreviation. It is "alt" for both "alternative" and "alternatives", and I am pretty sure I'll use concat and alt instead of writing the whole name. What about using found, switch, etc? Another option to use something related to the transformation itself and not to the "alternatives". For example `key(x, {a:a,b:b})`. The "key" is not the best here, but we can do "invert" or "map" or something similar. We can do "found", "matching".

**Comment by inf3rno** [comment]
Huhh. Map inversion is this: "{key:value}" -> "{value:key}". This is not entirely inversion. It is finding the key for the value on a map. It could be `key(x).on({a:a,b:b})` or `map({a:a,b:b}).key(x)`. Another option that we use a prefix object for mapping. E.g. `patterns.mapping.alternatives(x, {a:a,b:b})` vs `x = patterns.alternatives(a,b)`.

**Comment by inf3rno** [comment]
It could be `name(x).from.alternatives({a:a,b:b})` too.

**Comment by inf3rno** [comment]
I think we need to think it through what we are doing from the pattern matching context. We have a capturing group "x", which consist of different alternatives like "a" and "b". We need to decide which alternative matched. So the proper word would be `matchingAlternative(x) -> b`. On top of that we map that alternative to a string, so it won't depend on the patterns we use.

**Comment by inf3rno** [comment]
I think this can be closed. I have the basic idea and I have a list of features I need to support. I'll move that list to separate threads.
